[MLP-197][MLP-198][MLP-199][MLP-200] Unit Test for Role end points

GustavoAlMena opened the pull request:
## Background
Unit test for Role CRUD API methods 
## Changes done
- Add XUnit framework
- Add Unit Test for Role CRUD operations
## Notes
The nuget packages sometimes not be restored immediately for this case just rebuild the solution

GustavoAlMena committed:
[MLP-197][MLP-198][MLP-199][MLP-200] Modify unit test methods for Role Service

martinkeimel commented on P202.Training.Test/RoleServiceTest.cs:
```diff
@@ -0,0 +1,88 @@
+        public void RoleService_CreateData()
+        {
+            var newRole = new Role();
+            var mockpersonRepository = new Mock<IRoleRepository>();
```
why is it called `mockpersonRepository` if it's the mock of the `RoleRepository`? Perhaps you should rename this variable

GustavoAlMena replied:
Yes, it is a mistake. I copy/paste an example. I'll rename this variable as you suggest

martinkeimel commented on P202.Training.Test/RoleServiceTest.cs:
```diff
@@ -0,0 +1,88 @@
+            var newRole = new Role();
+            var mockpersonRepository = new Mock<IRoleRepository>();
+            mockpersonRepository.Setup(x => x.CreateRole(newRole)); 
+            mockpersonRepository.Object.CreateRole(newRole);
```
For the class name it seems your intention is to test the `RoleService` but you're never calling any of it's methods. You should instantiate a `RoleService` passing the mock as a parameter and then test it's methods. Something like this:
```
var roleService = new RoleService(mockRoleRepository.Object);
roleService.CreateRole(...);

// Assertions
```

DanielAltamirano replied:
I'm not sure we should have UT for the repository at all. If we wanted to test it, maybe they should be integration tests, and thus we shouldn't mock the repository methods.

GustavoAlMena replied:
@martinkeimel  I'll apply the solution that you suggest.

alcabelu committed:
[MLP 205]  Create page for listing existing roles

Add folder Role
Add file roleList.html in folder Role
Add file roleList.js in folder Role
Add file roleService.js in folder service
Add Code in routes.js for Role
Add functionality For Page for RoleList

martinkeimel reviewed (commented):
I left a couple of comments that applies to all the tests. Let me know if they are not clear

DanielAltamirano commented on P202.Training.Test/RoleServiceTest.cs:
```diff
@@ -0,0 +1,88 @@
+            // Assert
+            var model = Assert.IsType<List<Role>>(result);
+            Assert.Equal(2, model.Count);
+        }
```
This test has the right idea about different sections for setup/act/assert. Could you replicate that on the rest of the tests?

GustavoAlMena replied:
For sure @DanielAltamirano . I 'll replicate that on the rest of the test as you suggest

DanielAltamirano commented on P202.Training.Test/RoleServiceTest.cs:
```diff
@@ -0,0 +1,88 @@
+            mockpersonRepository.Setup(x => x.DeleteRole(deleteRole));
+            mockpersonRepository.Object.DeleteRole(deleteRole);
+            mockpersonRepository.Verify(x => x.DeleteRole(deleteRole), Times.Once());
+        }
```
What are we testing here?

GustavoAlMena replied:
I'm modifying this code because it's incomplete

DanielAltamirano commented on P202.Training.Test/RoleServiceTest.cs:
```diff
@@ -0,0 +1,128 @@
+        {
+            // Arrange
+            Mock<IRoleRepository> roleRepository = new Mock<IRoleRepository>();
+            Mock<IMapper> mappingService = new Mock<IMapper>();
```
Let's use `var` for these

DanielAltamirano replied:
Also, I'd add "Mock" as a suffix for those variables, so we remember those are not the real objects.

GustavoAlMena replied:
Done!

DanielAltamirano commented on P202.Training.Test/RoleServiceTest.cs:
```diff
@@ -0,0 +1,128 @@
+                Name = "Admin"
+            };
+
+            roleRepository.Setup(r => r.CreateRole(newRole));           
```
There are lots of spaces added at the end of the line, please check those are removed everywhere

GustavoAlMena replied:
Fixed!

DanielAltamirano commented on P202.Training.Test/RoleServiceTest.cs:
```diff
@@ -0,0 +1,128 @@
+            roleRepository.Setup(r => r.CreateRole(newRole));           
+
+            mappingService.Setup(m => m.Map<Role, Role>(It.IsAny<Role>()))
+                            .Returns(newRole);          
```
In this line you use `It.IsAny<Role>()` for the Setup, but not in the previous one. Let's try and be consistent in that regard. I'd use `It.IsAny` for all scenarios.

GustavoAlMena replied:
Yes, you are right. I'll use `It.IsAny`

DanielAltamirano commented on P202.Training.Test/RoleServiceTest.cs:
```diff
@@ -0,0 +1,128 @@
+                            .Returns(newRole);          
+
+            // Act
+            roleRepository.Object.CreateRole(newRole);
```
Really Minor: I'd assign `roleRepository.Object` to a variable and use it instead of referencing the internal property `Object` everytime.

GustavoAlMena replied:
You are right! Done!

alcabelu committed:
Made Changes for comments of gmsoft about blank lines in HTML Ctr+L, and console log for values of call WebService

DanielAltamirano commented on P202-Training.Domain/IMappingService.cs:
```diff
@@ -0,0 +1,13 @@
+{
+    public interface IMappingService
+    {
+        TDest Map<TSrc, TDest>(TSrc source) where TDest : class;
```
I think we need to enforce that `TSrc` and `TDest` cannot be the same class. Otherwise you might end up having service methods that receive a class to the domain and map it to the same class, which is what happened in the service classes in here. Also the tests have that issue. We need to fix that.

I'd have a base class for DTOs and a base class for Domain entities. That way you can define two methods. One for mapping a DTO to a Domain entity, and viceversa.

GustavoAlMena replied:
In this point, I gonna need help

DanielAltamirano replied:
No problem. Let's discuss after the meeting.

DanielAltamirano commented on P202.Training.Test/RoleServiceTest.cs:
```diff
@@ -0,0 +1,128 @@
+            //Assert that the Add method was called once
+            roleRepository.Verify(x => x.CreateRole(newRole), Times.Once());
+
+            Role expectedRole = roleService.ReadRole(1);
```
Let's use var

GustavoAlMena replied:
Done!

DanielAltamirano commented on P202.Training.Test/RoleServiceTest.cs:
```diff
@@ -0,0 +1,128 @@
+            var role = new Role { Id = 3 };
+            var mockRoleRepository = new Mock<IRoleRepository>();
+            mockRoleRepository.Setup(x => x.GetRole(3)).Returns(role);
+            mockRoleRepository.Object.GetRole(3);
```
If `GetRole` is not being used in the service because it's not implemented, what's the point of setting up the mock for the repository? It forces you to call the method in the test just to have something to assert. I think this whole test is moot until the method gets implemented for real.

GustavoAlMena replied:
Done! I've implemented the service as you suggest

DanielAltamirano commented on P202.Training.Test/RoleServiceTest.cs:
```diff
@@ -0,0 +1,128 @@
+        }
+
+        [Fact]
+        public void RoleService_ListOfData()
```
The Role service is nowhere in this test. What are we testing here then?

DanielAltamirano commented:
Please review the naming convention for tests and change the tests accordingly: https://github.com/MakingSense/code-style-guides/tree/master/CSharp#testing-code

GustavoAlMena commented on P202.Training.Test/RoleServiceTest.cs:
```diff
@@ -0,0 +1,128 @@
+    public class RoleServiceTest
+    {
+        [Fact]
+        public void RoleService_CreateData()
```
@DanielAltamirano  According to the naming convention for tests, what do you think about `CreateRole_PersistsChange`?

DanielAltamirano replied:
Sounds good!

GustavoAlMena replied:
Great. Thank you!

DanielAltamirano commented on P202-Training.Domain/IMappingService.cs:
```diff
@@ -0,0 +1,39 @@
+        /// </summary>
+        /// <typeparam name="TSrc"></typeparam>
+        /// <typeparam name="TDest"></typeparam>
+        /// <param name="source"></param>
```
If we're going to add headers, remember to add descriptions for each parameter.

DanielAltamirano replied:
Shouldn't this method be removed? Are we using it if we have the other two?

martinkeimel commented on P202-Training.Domain/MappingService.cs:
```diff
@@ -0,0 +1,28 @@
+
+namespace P202.Training.Domain
+{
+    public class MappingService : IMappingService
```
Is this class being used? Which are the classes that consume it? I'm not seeing any service having an `IMappingService` injected

GustavoAlMena replied:
No, this class is not necessary. I gonna remove this class.

martinkeimel commented on P202.Training.Test/RoleServiceTest.cs:
```diff
@@ -0,0 +1,139 @@
+
+            // Act
+            var roleRepositoryMockObject = roleRepositoryMock.Object;
+            roleRepositoryMockObject.CreateRole(newRole);
```
I'm almost sure this line is not needed. You shouldn't be calling mock methods from the UT. Just performing setups and verifications

GustavoAlMena replied:
Done!

martinkeimel commented on P202.Training.Test/RoleServiceTest.cs:
```diff
@@ -0,0 +1,139 @@
+        {
+            // Arrange
+            var roleRepositoryMock = new Mock<IRoleRepository>();
+            var mappingServiceMock = new Mock<IMapper>();
```
minor: rename to `mapperMock` so it's easier to identify which class we are mocking

GustavoAlMena replied:
Done!

martinkeimel reviewed (approved):
Good job!

DanielAltamirano commented:
I see the UT fail most of them (except for `DeleteRole_PersistsChange`). Do they pass in your environment, @gmsoft?

gmsoft commented:
@DanielAltamirano  Regarding "UT fail" I can see that I have the same situation in my environment. The problem is the following: 
```
var roleService = new RoleService(roleRepositoryMockObject, mappingServiceMockObject);
var expectedRole = roleService.ReadRole(1);
```
`expectedRole` returns `null` and I don't realize why. Could you help me?
